packages/loopring_v3/contracts/core/impl/libexchange/ExchangeDeposits.sol

```diff
@@ -46,26 +47,53 @@ library ExchangeDeposits
 
         uint16 tokenID = S.getTokenID(tokenAddress);
 
-        // Transfer the tokens to this contract
-        uint96 amountDeposited = S.depositContract.deposit{value: msg.value}(
-            from,
-            tokenAddress,
-            amount,
-            extraData
-        );
+        uint96 amountDeposited = amount;
+        if (flash) {
+            require(msg.value == 0, "INVALID_FLASH_DEPOSIT");
+            S.amountFlashMinted[tokenAddress] = S.amountFlashMinted[tokenAddress].add(amount);
+        } else {
+            // Transfer the tokens to this contract
+            amountDeposited = S.depositContract.deposit{value: msg.value}(
+                from,
+                tokenAddress,
+                amount,
+                extraData
+            );
+
+            emit DepositRequested(
+                from,
+                to,
+                tokenAddress,
+                tokenID,
+                amountDeposited
+            );
+        }
 
         // Add the amount to the deposit request and reset the time the operator has to process it
         ExchangeData.Deposit memory _deposit = S.pendingDeposits[to][tokenID];
         _deposit.timestamp = uint64(block.timestamp);
         _deposit.amount = _deposit.amount.add(amountDeposited);
         S.pendingDeposits[to][tokenID] = _deposit;
+    }
 
-        emit DepositRequested(
+    function repayFlashMint(
+        ExchangeData.State storage S,
+        address from,
+        address tokenAddress,
+        uint96  amount,
+        bytes   memory extraData
+        )
+        public
+    {
+        // Transfer the tokens to this contract
+        uint96 amountDeposited = S.depositContract.deposit{value: msg.value}(
             from,
-            to,
             tokenAddress,
-            tokenID,
-            amountDeposited
+            amount,
+            extraData
         );
+
+        // Paid back
+        S.amountFlashMinted[tokenAddress] = S.amountFlashMinted[tokenAddress].sub(amountDeposited);
```

Comment:
should we allow over-payment?

Reply:
The amount flash minted and repaid is completely deterministic so I don't see a reason why they would ever be different. Unless as some kind of fee, but because we handle the handle the on-chain logic for flash mints that doesn't seem to make a lot of sense, we would just ask for a fee on L2.